File: java/src/main/java/io/fivexer/sdk/model/WorkflowDefinition.java
package io.fivexer.sdk.model;

import java.util.List;
import java.util.Map;

/**
 * A stored workflow graph.
 *
 * <p>Populated by Gson via field reflection and exposed read-only; callers receive instances
 * from the client and never construct them.
 */
public class WorkflowDefinition {
    private String id;
    private String name;
    private int version;
    private String initialStepId;
    private List<WorkflowStep> steps;
    private Long defaultTimeoutMs;
    private Map<String, Object> metadata;

    public String getId() { return id; }
    public String getName() { return name; }
    public int getVersion() { return version; }
    public String getInitialStepId() { return initialStepId; }
    public List<WorkflowStep> getSteps() { return steps; }
    public Long getDefaultTimeoutMs() { return defaultTimeoutMs; }
    public Map<String, Object> getMetadata() { return metadata; }
}

File: php/src/Model/WorkflowDefinition.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Model;

use Fivexer\SDK\Internal\Json;

/**
 * A stored workflow graph.
 *
 * Response model: built from a decoded JSON array and read-only thereafter.
 */
final class WorkflowDefinition
{
    public function __construct(
        public readonly string $id,
        public readonly string $name,
        public readonly int $version,
        public readonly string $initialStepId,
        /** @var list<WorkflowStep> */
        public readonly array $steps,
        public readonly ?int $defaultTimeoutMs,
        /** @var array<string, mixed>|null */
        public readonly ?array $metadata,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     * @internal
     */
    public static function fromArray(array $data): self
    {
        return new self(
            id: (string) ($data['id'] ?? ''),
            name: (string) ($data['name'] ?? ''),
            version: (int) ($data['version'] ?? 0),
            initialStepId: (string) ($data['initialStepId'] ?? ''),
            steps: Json::parseEach($data, 'steps', [WorkflowStep::class, 'fromArray']),
            defaultTimeoutMs: isset($data['defaultTimeoutMs']) ? (int) $data['defaultTimeoutMs'] : null,
            metadata: $data['metadata'] ?? null,
        );
    }
}
